### PR TITLE
created policies with Pundit and updated CRUD features

### PR DESCRIPTION
**Pull request by APatel-AI**

### CODE REVIEW

File: app/controllers/follow_requests_controller.rb [hunk 1]
```diff
@@ -67,4 +72,9 @@ def set_follow_request
     def follow_request_params
       params.require(:follow_request).permit(:recipient_id, :sender_id, :status)
     end
+
+    def authorize_follow_request
```
Comment (dylansarikas) [inline app/controllers/follow_requests_controller.rb:76]: nice job creating this method and calling it in a before action

**Review by dylansarikas — COMMENTED** [review]
looks good arpan!

**Review by dylansarikas — APPROVED** [review]

**Review by dylansarikas — APPROVED** [review]

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,12 +1,22 @@
 class ApplicationController < ActionController::Base
+  include Pundit
   before_action :authenticate_user!
-
   before_action :configure_permitted_parameters, if: :devise_controller?
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
 
-  protected
+  after_action :verify_authorized, unless: :devise_controller?
```
Comment (ajmompr) [inline app/controllers/application_controller.rb:7]: good job accounting for devise

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -1,30 +1,29 @@
 class CommentsController < ApplicationController
   before_action :set_comment, only: %i[ show edit update destroy ]
```
Comment (ajmompr) [inline app/controllers/comments_controller.rb]: missing before_action to authorize comment policy. How are you ensuring a user is authorized to comment?

File: app/controllers/follow_requests_controller.rb [hunk 4]
```diff
@@ -1,5 +1,6 @@
 class FollowRequestsController < ApplicationController
   before_action :set_follow_request, only: %i[ show edit update destroy ]
+  before_action :authorize_follow_request, only: [:show, :edit, :update, :destroy]
```
Comment (ajmompr) [inline app/controllers/follow_requests_controller.rb:3]: good

File: app/controllers/photos_controller.rb [hunk 5]
```diff
@@ -1,19 +1,27 @@
 class PhotosController < ApplicationController
   before_action :set_photo, only: %i[ show edit update destroy ]
   before_action :ensure_current_user_is_owner, only: [:destroy, :update, :edit]
+  before_action :authorize_photo, only: [:show, :edit, :update, :destroy]
```
Comment (ajmompr) [inline app/controllers/photos_controller.rb:3]: good

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -1,13 +1,17 @@
 class UsersController < ApplicationController
   before_action :set_user, only: %i[ show liked feed followers following discover ]
+  before_action {authorize @user || User}
```
Comment (ajmompr) [inline app/controllers/users_controller.rb:3]: great! You can use similar before actions across all of your controllers. This is more concise.

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -1,13 +1,17 @@
...
         @user = User.find_by!(username: params.fetch(:username))
       else
         @user = current_user
       end
     end
-end
\ No newline at end of file
+
+    def authorize_user
+      authorize @user
```
Comment (ajmompr) [inline app/controllers/users_controller.rb]: if calling before_action {authorize @user || User}, authorize @ user becomes redundant

File: app/policies/like_policy.rb
Comment (ajmompr) [inline app/policies/like_policy.rb:1]: not needed for this assignment

File: app/policies/user_policy.rb
Comment (ajmompr) [inline app/policies/user_policy.rb:1]: perfect

File: app/views/users/show.html.erb [hunk 8]
```diff
@@ -4,7 +4,7 @@
   </div>
 </div>
 
-<% if current_user == @user || !@user.private? || current_user.leaders.include?(@user) %>
+<% if policy(@user).show? %>
```
Comment (ajmompr) [inline app/views/users/show.html.erb]: should be calling show_photos?

**Review by ajmompr — CHANGES_REQUESTED** [review]
Fix the policy that you are calling in users#show. Add missing before action to your comments controller

File: app/controllers/comments_controller.rb [hunk 9]
```diff
@@ -1,30 +1,31 @@
 class CommentsController < ApplicationController
-  before_action :set_comment, only: %i[ show edit update destroy ]
-  before_action :is_an_authorized_user, only: [:destroy, :create]
+  before_action :set_comment, only: %i[ show edit update destroy]
+  before_action {authorize @comment || Comment}
```
Comment (ajmompr) [inline app/controllers/comments_controller.rb:3]: good

**Review by ajmompr — APPROVED** [review]

**Review by FadiBaker92 — COMMENTED** [review]
code looks clean. Good job Arpan!

**Pull request merged**